File: apps/ios/Sources/Litter/Models/TerminalOutputBuffer.swift
import Foundation

struct TerminalOutputBuffer {
    private let maxCharacters: Int
    private let maxPendingCharacters: Int
    private(set) var text = ""
    private var pending = ""

    init(maxCharacters: Int = 64_000, maxPendingCharacters: Int = 64_000) {
        self.maxCharacters = maxCharacters
        self.maxPendingCharacters = maxPendingCharacters
    }

    var hasPendingOutput: Bool {
        !pending.isEmpty
    }

    mutating func append(_ data: Data) {
        guard !data.isEmpty else { return }
        pending += String(decoding: data, as: UTF8.self)
        trimPendingIfNeeded()
    }

    mutating func flush() -> Bool {
        guard !pending.isEmpty else { return false }
        text = String((text + pending).suffix(maxCharacters))
        pending.removeAll(keepingCapacity: true)
        return true
    }

    mutating func clear() {
        text.removeAll(keepingCapacity: true)
        pending.removeAll(keepingCapacity: true)
    }

    private mutating func trimPendingIfNeeded() {
        guard pending.count > maxPendingCharacters else { return }
        pending = String(pending.suffix(maxPendingCharacters))
    }
}

File: apps/android/app/src/main/java/com/litter/android/state/TerminalOutputBuffer.kt
package com.litter.android.state

internal class TerminalOutputBuffer(
    private val maxChars: Int = 64_000,
    private val maxPendingChars: Int = 64_000,
) {
    var text: String = ""
        private set

    private val pending = StringBuilder()

    val hasPending: Boolean
        get() = pending.isNotEmpty()

    fun append(data: ByteArray) {
        if (data.isEmpty()) return
        pending.append(data.toString(Charsets.UTF_8))
        trimPendingIfNeeded()
    }

    fun flush(): Boolean {
        if (pending.isEmpty()) return false
        text = (text + pending.toString()).takeLast(maxChars)
        pending.clear()
        return true
    }

    fun clear() {
        text = ""
        pending.clear()
    }

    private fun trimPendingIfNeeded() {
        if (pending.length <= maxPendingChars) return
        val trimmed = pending.takeLast(maxPendingChars).toString()
        pending.clear()
        pending.append(trimmed)
    }
}
